File: cs/sci_compute_lag/trangerstein_book_extras/ode/euler.c
#include <stdio.h> /* for FILE, scanf, printf, fprintf */
#include <stdlib.h> /* for EXIT_SUCCESS */

/*main program must have the following argument list and return value*/
int main(int argc,char **argv) {
/*must declare all variables first*/
  int i=0,nsteps=0; 
  float dt=1.,rate=1.,t=0.,tmax=1.,u=1.;
  FILE *out;

/*function definitions*/
  float f(float u,float t) { return rate*u; }
  float euler(float u,float t,float dt) { return u+dt*f(u,t); }

/*read input from stdin*/
  printf("enter rate\n");
  scanf("%f",&rate);
  printf("enter initial u\n");
  scanf("%f",&u);
  printf("enter tmax\n");
  scanf("%f",&tmax);
  printf("enter number steps\n");
  scanf("%d",&nsteps);
  printf("rate = %f u = %f\n",rate,u);
  printf("tmax = %f number steps = %d\n",tmax,nsteps);

/*apply forward-Euler method*/
  t=0.;
  dt=tmax/((float) nsteps);
  out=fopen("coutput","w");
  fprintf(out,"%f %f\n",t,u);
  for (i=1;i<=nsteps;i++) {
    u=euler(u,t,dt);
    t+=dt;
    fprintf(out,"%f %f\n",t,u);
  }
  fclose(out);
  return EXIT_SUCCESS;
}

File: cs/sci_compute_lag/trangerstein_book_extras/ode/euler.f
      program main
c declare variables:
      integer i,nsteps
      real dt,f,rate,t,tmax,u

c arithmetic statement functions:
      f(u,t)=rate*u
      euler(t,u,dt)=u+dt*f(u,t)

c read input:
      print *, "enter rate"
      read *, rate
      print *, "enter initial u"
      read *, u
      print *, "enter tmax"
      read *, tmax
      print *, "enter number steps"
      read *, nsteps
      print *, "rate = ",rate," u = ",u
      print *, " tmax = ",tmax," number steps = ",nsteps

c apply forward-Euler method:
      t=0.
      dt=tmax/real(nsteps)
      write(10,*) t,u
      do i=1,nsteps
        u=euler(t,u,dt)
        t=t+dt
        write(10,*) t,u
      enddo

      stop
      end
